### PR TITLE
boot: prune some bootchains that should not happen

### PR DESCRIPTION
**Pull request by valentindavid**
We are supposed to update on disk from the back of the boot chain to the beginning. That means and old shim should not be expected to install a new grub.

We can reduce the list of possible boot chains we give to secboot. When we will switch to secboot master, secboot will fail if it sees boot chains that cannot work. So we should make sure not to pass them.

There is still a bug where we update the seed partition before boot partition. But here we assume the bug is fixed.

### CODE REVIEW

File: boot/bootchain.go [hunk 1]
```diff
@@ -270,7 +270,20 @@ func bootAssetsToLoadChains(assets []bootAsset, kernelBootFile bootloader.BootFi
 		return nil, fmt.Errorf("internal error: no bootloader name for boot asset role %q", thisAsset.Role)
 	}
 	var chains []*secboot.LoadChain
-	for _, hash := range thisAsset.Hashes {
+
+	for i, hash := range thisAsset.Hashes {
+		// There should be 1 or 2 assets, and their position has a meaning.
+		// See TrustedAssetsUpdateObserver.observeUpdate
+		if i == 0 {
```
Comment (alfonsosanchezbeato) [inline boot/bootchain.go:277]: It looks like i==0 is the old asset and i==1 the new one, but please add a comment to clarify, also worth a longer comment saying basically what you say in the commit, I think it is important enough so we keep that in mind while reading the code.

**Review by alfonsosanchezbeato — COMMENTED** [review]
Some minor comments

File: boot/assets.go [hunk 2]
```diff
@@ -600,6 +600,11 @@ func (o *TrustedAssetsUpdateObserver) observeUpdate(bl bootloader.Bootloader, re
 			// during an update; more entries indicates that the
 			// same asset name is used multiple times with different
 			// content
+
+			// The order is important. Changing it would
+			// change assumptions in
+			// bootAssetsToLoadChains
+
```
[the remark this answers is not in the document]
Comment (valentindavid) [reply boot/assets.go]: I agree. I moved it.

File: boot/bootchain.go [hunk 3]
```diff
@@ -270,7 +270,20 @@ func bootAssetsToLoadChains(assets []bootAsset, kernelBootFile bootloader.BootFi
...
+			// First asset is currently installed asset
+			if len(thisAsset.Hashes) == 2 && expectNext {
+				continue
+			}
+		} else if i != 1 {
```
[the remark this answers is not in the document]
Comment (valentindavid) [reply boot/bootchain.go]: I have changed the order of conditions. So that it is `i == 0`, `i == 1`, then any other.

File: boot/bootchain.go [hunk 4]
```diff
@@ -275,11 +275,16 @@ func bootAssetsToLoadChains(assets []bootAsset, kernelBootFile bootloader.BootFi
 		// There should be 1 or 2 assets, and their position has a meaning.
 		// See TrustedAssetsUpdateObserver.observeUpdate
 		if i == 0 {
-			// First asset is currently installed asset
+			// i == 0 means currently installed asset
 			if len(thisAsset.Hashes) == 2 && expectNext {
+				fmt.Fprintf(os.Stderr, "Ignoring %s %s hash %s\n", thisAsset.Role, thisAsset.Name, hash)
```
Comment (alfonsosanchezbeato) [inline boot/bootchain.go:280]: Why these prints to stderr? Shouldn't this use logger.Noticef?

**Review by alfonsosanchezbeato — COMMENTED** [review]
A comment about printing. Also, I think that a comment like:
"We do not calculate some boot chains because they are impossible as when we update assets we write first the binaries that are used later, that is, if updating both shim and grub, the new grub is copied first to the disk, so booting from the new shim to the old grub is not possible. This is controlled by expectNext, that tells us that the previous step in the chain is from a new asset."
would be nice to have around.

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, thanks for the changes!

File: boot/bootchain.go [hunk 5]
```diff
@@ -257,7 +257,13 @@ func predictableBootChainsEqualForReseal(pb1, pb2 predictableBootChains) bootCha
...
 // assets sequence. At the end of each chain, adds an entry for the kernel boot
 // file.
-func bootAssetsToLoadChains(assets []bootAsset, kernelBootFile bootloader.BootFile, roleToBlName map[bootloader.Role]string) ([]*secboot.LoadChain, error) {
+// We do not calculate some boot chains because they are impossible as
+// when we update assets we write first the binaries that are used
+// later, that is, if updating both shim and grub, the new grub is
+// copied first to the disk, so booting from the new shim to the old
+// grub is not possible. This is controlled by expectNext, that tells
+// us that the previous step in the chain is from a new asset.
+func bootAssetsToLoadChains(assets []bootAsset, kernelBootFile bootloader.BootFile, roleToBlName map[bootloader.Role]string, expectNext bool) ([]*secboot.LoadChain, error) {
```
Comment (pedronis) [inline boot/bootchain.go:266]: wouldn't expectNew be clearer?

File: boot/bootchain.go [hunk 6]
```diff
@@ -270,7 +276,22 @@ func bootAssetsToLoadChains(assets []bootAsset, kernelBootFile bootloader.BootFi
...
 	var chains []*secboot.LoadChain
-	for _, hash := range thisAsset.Hashes {
+
+	for i, hash := range thisAsset.Hashes {
+		// There should be 1 or 2 assets, and their position has a meaning.
+		// See TrustedAssetsUpdateObserver.observeUpdate
+		if i == 0 {
+			// i == 0 means currently installed asset
+			if len(thisAsset.Hashes) == 2 && expectNext {
+				continue
```
Comment (pedronis) [inline boot/bootchain.go:286]: // we do not expect it to use as we have new assets earlier in the chain

File: boot/seal.go [hunk 7]
```diff
@@ -861,7 +861,7 @@ func sealKeyModelParams(pbc predictableBootChains, roleToBlName map[bootloader.R
 	for _, bc := range pbc {
 		modelForSealing := bc.modelForSealing()
 		modelID := modelUniqueID(modelForSealing)
-		loadChains, err := bootAssetsToLoadChains(bc.AssetChain, bc.kernelBootFile, roleToBlName)
+		loadChains, err := bootAssetsToLoadChains(bc.AssetChain, bc.kernelBootFile, roleToBlName, false)
```
Comment (pedronis) [inline boot/seal.go:864]: we tend to write this kind of code with constants: maybe:

// expect to start from both old and new assets
const expectNew = false
..., expectNew)

File: boot/bootchain.go [hunk 8]
```diff
@@ -270,7 +276,22 @@ func bootAssetsToLoadChains(assets []bootAsset, kernelBootFile bootloader.BootFi
...
+			}
+		} else if i == 1 {
+			// i == 1 means new asset
+		} else {
+			// If there is a second asset, it is the next asset to be installed
+			return nil, fmt.Errorf("did not expect more than 2 hashes for %s", thisAsset.Name)
```
Comment (pedronis) [inline boot/bootchain.go:292]: internal error: ...

**Review by pedronis — COMMENTED** [review]
thanks, some small comments

**Review by pedronis — APPROVED** [review]
thanks

**Pull request closed without merging**